File: lec5/ABB140.m


%% Section1 
%            alpha      a        d     modified D&H
%                                      Because we use Craig            

T= [1 0 0 0;
      0 1 0 0;
      0 0 1 352;
      0 0 0 1];

L1=Link('d',  0,'a',  0,'alpha',  0,       'modified');
L2=Link('d',  0,'a', 70,'alpha',-90*pi/180,'offset',-90*pi/180,'modified');
L3=Link('d',  0,'a',360,'alpha',  0,       'modified');
L4=Link('d',380,'a',  0,'alpha',-90*pi/180,'modified');
L5=Link('d',  0,'a',  0,'alpha', 90*pi/180,'modified');
L6=Link('d',  0,'a',  0,'alpha',-90*pi/180,'modified');
 
ABB140x=SerialLink([L1 L2 L3 L4 L5 L6], 'name', 'ABB');
 
q=[10 20 30 40 50 60]*pi/180;

T06=cast(ABB140x.fkine(q), 'like', T)  

RZYX = tr2rpy(T06,'deg','zyx')

    

%% Section 2 Move base frame

TB0= [1 0 0 0;
      0 1 0 0;
      0 0 1 352;
      0 0 0 1];

TB6 = TB0*T06

RZYX = tr2rpy(TB6,'deg','zyx')

%% Step 3 Move wrist frame

T6W= [-1  0  0  0;
      0  -1  0  0;
      0  0  1 65;
      0  0  0  1];

TBW=TB0*T06*T6W  
ZYX = tr2rpy(TBW,'deg','zyx')
 
%% step 4 With tool data

TWT=transl(-4,0,371.300)*troty(45*pi/180);
TBT=TB0*T06*T6W*TWT
ZYX = tr2rpy(TBT,'deg','zyx')

File: lec5/ABB140.py
from math import pi
from roboticstoolbox import DHRobot, RevoluteMDH
import numpy as np
from spatialmath import SE3

# Section 1 - Denavit-Hartenberg Parameters

# Homogeneous Transformation matrix for the base
T = np.array([[1, 0, 0, 0],
              [0, 1, 0, 0],
              [0, 0, 1, 352],
              [0, 0, 0, 1]])

# Define the links using Denavit-Hartenberg parameters
L = [
	RevoluteMDH(alpha =    0,        a=0,   d=0),
    RevoluteMDH(alpha = -90*pi/180,  a=70,  d=0,  offset=-90*pi/180),
    RevoluteMDH(alpha =     0,       a=360, d=0),
    RevoluteMDH(alpha=-90*pi/180,    a=0,   d=380),
    RevoluteMDH(alpha=90*pi/180,     a=0,   d=0),
    RevoluteMDH(alpha=-90*pi/180,    a=0,   d=0, ),
]

# Create the robot model
ABB140 = DHRobot(L, name='ABB')

# Define joint angles in radians
q = np.radians([10, 20, 30, 40, 50, 60])

# Forward kinematics to get the transformation matrix
T06 = ABB140.fkine(q)

# Extract Roll, Pitch, Yaw from T06 (ZYX convention)
RZYX = T06.rpy(order='zyx',unit='deg')

print("Rotation Angles (ZYX) for Section 1:", RZYX)

# Section 2 - Move base frame

TB0 = SE3([[1, 0, 0, 0],
           [0, 1, 0, 0],
           [0, 0, 1, 352],
           [0, 0, 0, 1]])

# Compute TB6 by multiplying TB0 with T06
TB6 = TB0 * T06

# Extract Roll, Pitch, Yaw from TB6
RZYX = TB6.rpy(order='zyx', unit='deg')

print("Rotation Angles (ZYX) for Section 2:", RZYX)

# Section 3 - Move wrist frame

T6W = SE3([[-1, 0, 0, 0],
           [0, -1, 0, 0],
           [0, 0, 1, 65],
           [0, 0, 0, 1]])

# Compute TBW by multiplying TB0, T06, and T6W
TBW = TB0 * T06 * T6W
print(TBW)
# Extract Roll, Pitch, Yaw from TBW
ZYX = TBW.rpy(order="zyx", unit="deg")

print("Rotation Angles (ZYX) for Section 3:", ZYX)

# Section 4 - With Tool Data

TWT = SE3.Trans(-4, 0, 371.300) * SE3.Ry(45,unit='deg')

# Compute TBT by multiplying TB0, T06, T6W, and TWT
TBT = TB0 * T06 * T6W * TWT

# Extract Roll, Pitch, Yaw from TBT
ZYX = TBT.rpy(order='zyx', unit='deg')

print("Rotation Angles (ZYX) for Section 4:", ZYX)
print(TBT)
